### PR TITLE
add new erc20

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -240,6 +243,15 @@ func (m *manager) insertBlock(ctx context.Context, dbTx *gorm.DB, block *types.B
 		return err
 	}
 
+	// Init new erc20 tokens if existed
+	newTokens, err := m.initNewERC20(ctx, accountStore, subsStore, blockNumber, m.balancer)
```
[the remark this answers is not in the document]
Comment (kimmylin) [reply store/store.go]: i think maybe it's worth putting this in another DB transaction before insertBlock.

File: store/subscription/subscription.go [hunk 2]
```diff
@@ -138,10 +139,16 @@ func (t *store) Reset(from, to int64) error {
 }
 
 func (t *store) FindByGroup(groupID int64, query *model.QueryParameters) ([]*model.Subscription, uint64, error) {
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/subscription/subscription.go]: OK

File: store/subscription/subscription.go [hunk 3]
```diff
@@ -138,10 +139,16 @@ func (t *store) Reset(from, to int64) error {
...
-	filter := &model.Subscription{
+	return t.find(t.db.Model(&model.Subscription{}).Where(&model.Subscription{
 		Group: groupID,
-	}
-	db := t.db.Model(&model.Subscription{}).Where(filter)
+	}), query)
+}
+
+func (t *store) ListOldSubscriptions(query *model.QueryParameters) ([]*model.Subscription, uint64, error) {
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/subscription/subscription.go]: OK

File: store/subscription/subscription_test.go [hunk 4]
```diff
@@ -289,6 +289,83 @@ var _ = Describe("Database Test", func() {
...
+					Limit:   1,
+					OrderBy: "created_at",
+					Order:   "asc",
+				})
+				Expect(err).Should(Succeed())
+				Expect(total).Should(Equal(uint64(2)))
+				Expect(len(result)).Should(Equal(1))
+				Expect(result[0].Group).Should(Equal(subs[1].Group))
+				Expect(result[0].Address).Should(Equal(subs[1].Address))
+			})
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/subscription/subscription_test.go]: OK

File: store/account/account.go [hunk 5]
```diff
@@ -83,12 +85,22 @@ func (t *store) FindERC20(address common.Address) (result *model.ERC20, err erro
...
 	return
 }
 
+func (t *store) ListNewERC20() (results []*model.ERC20, err error) {
+	results = []*model.ERC20{}
+	err = t.db.Where("block_number = 0").Find(&results).Error
+	return
+}
+
+func (t *store) SetERC20Block(code *model.ERC20) error {
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/account/account.go]: OK

File: store/new_erc20.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+			err = subStore.InsertTotalBalance(tb)
+			if err != nil {
+				log.Error("Failed to insert total balance", "err", err)
+				return nil, err
+			}
+		}
+	}
+
+	// Update erc20 tokens
+	for _, token := range newTokens {
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/new_erc20.go]: OK

File: store/new_erc20.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	// Return if there is no new ERC20
+	if len(list) == 0 {
+		return nil, nil
+	}
+
+	newTokens := make(map[gethCommon.Address]*model.ERC20, len(list))
+	for _, e := range list {
+		e.BlockNumber = blockNumber
+		newTokens[gethCommon.BytesToAddress(e.Address)] = e
+	}
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/new_erc20.go]: OK

File: store/new_erc20.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+			subMap[gethCommon.BytesToAddress(sub.Address)] = sub
+		}
+		// Contructs the requested balance map
+		contractsAddrs := make(map[gethCommon.Address]map[gethCommon.Address]struct{})
+		for _, token := range newTokens {
+			contractsAddrs[gethCommon.BytesToAddress(token.Address)] = make(map[gethCommon.Address]struct{})
+			for _, sub := range subs {
+				contractsAddrs[gethCommon.BytesToAddress(token.Address)][gethCommon.BytesToAddress(sub.Address)] = struct{}{}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/new_erc20.go]: OK

File: store/new_erc20.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+		results, err := balancer.BalanceOf(ctx, big.NewInt(blockNumber), contractsAddrs)
+		if err != nil {
+			logger.Error("Failed to get ERC20 balance", "len", len(contractsAddrs), "err", err)
+			return nil, err
+		}
+
+		// Update total balances
+		for contractAddr, addrs := range results {
+			for addr, balance := range addrs {
+				sub, ok := subMap[addr]
```
[the remark this answers is not in the document]
Comment (markya0616) [reply store/new_erc20.go]: But we cannot ensure whether the results from balancer are correct. Just add a protection.

File: store/store.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (markya0616) [reply store/store.go]: OK

File: store/new_erc20_test.go [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+		db.Delete(&model.Subscription{})
+		db.DropTable(model.Transfer{
+			Address: erc20s[0].Address,
+		})
+		db.DropTable(model.Transfer{
+			Address: erc20s[1].Address,
+		})
+	})
+
+	FIt("should be successful", func() {
```
Comment (kimmylin) [inline store/new_erc20_test.go]: undo `F`

File: store/account/account_test.go [hunk 11]
```diff
@@ -96,6 +96,75 @@ var _ = Describe("Account Database Test", func() {
 		})
 	})
 
+	It("ListOldERC20(), ListNewERC20(), SetERC20Block()", func() {
```
Comment (kimmylin) [inline store/account/account_test.go]: /s/SetERC20Block/BatchUpdateERC20BlockNumber/

File: store/new_erc20_test.go (hunk 10 above)
Comment (markya0616) [reply store/new_erc20_test.go]: OK

**Pull request merged**